input/kafkamdm/kafkamdm.go

```diff
@@ -162,19 +159,26 @@ func ConfigProcess(instance string) {
 	}
 
 	// initialize our offset metrics
-	partitionOffset = make(map[int32]*stats.Gauge64)
-	partitionLogSize = make(map[int32]*stats.Gauge64)
-	partitionLag = make(map[int32]*stats.Gauge64)
+	kafkaStats = make(map[int32]*KafkaStats)
 	for _, part := range partitions {
+		ks := KafkaStats{}
 		// metric input.kafka-mdm.partition.%d.offset is the current offset for the partition (%d) that we have consumed.
-		partitionOffset[part] = stats.NewGauge64(fmt.Sprintf("input.kafka-mdm.partition.%d.offset", part))
+		stats.NewGauge64Existing(fmt.Sprintf("input.kafka-mdm.partition.%d.offset", part), &ks.offset)
 		// metric input.kafka-mdm.partition.%d.log_size is the current size of the kafka partition (%d), aka the newest available offset.
-		partitionLogSize[part] = stats.NewGauge64(fmt.Sprintf("input.kafka-mdm.partition.%d.log_size", part))
+		stats.NewGauge64Existing(fmt.Sprintf("input.kafka-mdm.partition.%d.log_size", part), &ks.logSize)
 		// metric input.kafka-mdm.partition.%d.lag is how many messages (metrics) there are in the kafka partition (%d) that we have not yet consumed.
-		partitionLag[part] = stats.NewGauge64(fmt.Sprintf("input.kafka-mdm.partition.%d.lag", part))
+		stats.NewGauge64Existing(fmt.Sprintf("input.kafka-mdm.partition.%d.lag", part), &ks.lag)
```

Comment:
instead of adding this new stats.NewGauge64Existing method, wouldnt it be cleaner to just have a  `NewkafkaStats(prefix string) *KafkaStats` function.

eg,
```
kafkaStats[part] = NewKafkaStats(fmt.Sprintf("input.kafka-mdm.partition.%d", part))
...
func NewKafkaStats(prefix string) *KafkaStats {
  return &KafkaStats{
    offset: stats.NewGauge64(fmt.Sprintf("%s.offset", prefix)),
    logSize: stats.NewGauge64(fmt.Sprintf("%s.log_size", prefix)),
    lag: stats.NewGauge64(fmt.Sprintf("%s.lag", prefix)),
  }
}
```

Reply:
stats.NewGauge64Existing enables you to allocate on the callers terms (e.g. several gauges together in a known contiguous chunk of memory and without needing to track pointers to all of them separately), as stats.NewGauge64 allocates them individually - possibly in different memory locations - and forces you to refer to them as pointers.

The NewKafkaStats constructor function is a separate thing. Since there's only 1 place where we use this struct i was trying to avoid boilerplate and needless indirections

Reply:
sounds like an unnecessary optimization to me.

`stats.NewGauge64Existing` is a really odd looking function.  Why is the Gauge64 type so special that it has this but no other stats types do?

It looks to me like you just want to add exported functions to raintank/stats so you can add directly to the stats registry.

eg
```
func NewKafkaStats(prefix string) *KafkaStats {
  ks := KafkaStats{}
  stats.Add(fmt.Sprintf("%s.offset", prefix), &ks.offset)
  stats.Add(fmt.Sprintf("%s.log_size", prefix), &ks.log_size)
  stats.Add(fmt.Sprintf("%s.lag", prefix), &ks.lag)
  return &ks
}
```

Reply:
>  Why is the Gauge64 type so special that it has this but no other stats types do?

it was the only one that I needed at this point.

> It looks to me like you just want to add exported functions to raintank/stats 

assuming you mean metrictank/stats, exposing the registry makes sense.
I think I have another idea to clean this up, which is defining the KafkaStats type in the stats package.  We'll want the same stats tracking for notifierKafka as well so i think it makes sense. will give that a shot.